pallets/parachain-system/src/lib.rs

```diff
@@ -485,25 +485,38 @@ impl<T: Config> Module<T> {
 		downward_messages: Vec<InboundDownwardMessage>,
 	) -> Result<Weight, DispatchError> {
 		let dm_count = downward_messages.len() as u32;
-
 		let mut weight_used = 0;
 
-		// Reference fu to avoid the `move` capture.
-		let weight_used_mut_ref = &mut weight_used;
-		let result_mqc_head = LastDmqMqcHead::mutate(move |mqc| {
-			for downward_message in downward_messages {
-				mqc.extend_downward(&downward_message);
-				*weight_used_mut_ref += T::DownwardMessageHandlers::handle_downward_message(downward_message);
-			}
-			mqc.0
-		});
+		if dm_count != 0 {
+			let mut processed_count = 0;
 
-		// After hashing each message in the message queue chain submitted by the collator, we should
-		// arrive to the MQC head provided by the relay chain.
-		ensure!(
-			result_mqc_head == expected_dmq_mqc_head,
-			Error::<T>::DmpMqcMismatch
-		);
+			Self::deposit_event(RawEvent::DownwardMessagesReceived(dm_count));
+
+			// Reference fu to avoid the `move` capture.
+			let weight_used_ref = &mut weight_used;
+			let processed_count_ref = &mut processed_count;
+			let result_mqc_head = LastDmqMqcHead::mutate(move |mqc| {
+				for downward_message in downward_messages {
+					mqc.extend_downward(&downward_message);
+					*weight_used_ref += T::DownwardMessageHandlers::handle_downward_message(downward_message);
+					*processed_count_ref += 1;
+				}
+				mqc.0
+			});
+
+			Self::deposit_event(RawEvent::DownwardMessagesProcessed(
+				processed_count,
+				weight_used,
+				result_mqc_head.clone(),
+				expected_dmq_mqc_head.clone(),
+			));
+
+			// After hashing each message in the message queue chain submitted by the collator, we should
+			// arrive to the MQC head provided by the relay chain.
+			assert_eq!(result_mqc_head, expected_dmq_mqc_head);
```

Comment:
we could put this above the deposit event and then avoid the clone?

Reply:
Also we want this panic?

Reply:
yeah - it was relying on an indirect panic before. this just makes it explicit.

Reply:
also done in this PR: https://github.com/paritytech/cumulus/pull/413